### ISSUE TITLE
[feature request] Introduce torch.BoolTensor that convert nicely to numpy.bool array

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#4764]
I was surprised to find out it doesn't fall with an exception, but unfortunately the shapes are inconsistent.

```python
a = np.zeros((2, 4, 3))
b = np.ones((2, 4), dtype = np.bool)
c = torch.ones((2, 4)).eq(1)

print(a[b].shape)
# (8, 3)

print(a[c].shape)
# (2, 4, 4, 3)
```

Is seems because PyTorch doesn't surface BoolTensor, and casts ByteTensor to `np.uint8`. Workaround:

```python
a[c.numpy().astype(np.bool)]
```

**Comment by apaszke** [comment]
That's probably because `Byte` tensors get casted to `uint8` arrays, so what you end up doing is indexing instead of masking

**Comment by vadimkantorov** [comment]
You are right. This is probably not very critical, but are there plans to split ByteTensor and BoolTensor? I saw some mentions in the codebase, but if I understand well, it's there only for binding simplicity purposes.

**Comment by apaszke** [comment]
I don’t think we were planning to do that, but maybe at some point. BoolTensor could behave exactly like ByteTensor does, except that it converts to numpy correctly

**Comment by MahdiNazemi** [comment]
Another advantage of having `BoolTensor` is that it enables efficient implementation of binary neural networks. Currently, one should use a `ByteTensor` to implement such networks. As a result, different bit-wise operations such as AND, OR, and XOR are performed on eight bits rather than a single bit.

Because binary neural networks may have a better memory/energy efficiency, they are suitable for energy-constrained platforms like smartphones. Since ONNX supports Boolean tensors, it may be a good idea to introduce BoolTensor to allow conversion from PyTorch to ONNX, and later to Caffe2.

**Comment by fmassa** [comment]
Linking to issue https://github.com/pytorch/pytorch/issues/4219 , which also mentions bool tensors.

**Comment by vadimkantorov** [comment]
Fix incoming in https://github.com/pytorch/pytorch/pull/18505
